### ISSUE TITLE
Panics with 'implementation error, cannot get ref Int64 from Date64'

### ISSUE THREAD

**Issue by TheDan64** [ritchie46/polars#573]
#### Are you using Python or Rust?

Rust

#### What version of polars are you using?

Latest master, 04ec9ab4b66835c79cada2e4534c946a6c0361bb

#### What operating system are you using polars on?

Windows

#### Describe your bug.

Upgraded code to use latest polars, and it now panics instead of continuing to work correctly.

#### What are the steps to reproduce the behavior?

```Rust
fn foo(lf: LazyFrame, column_name: &str) -> LazyFrame {
    let maxima_lf = lf.clone().with_columns(vec![
        col(column_name).shift(1).lt(col(column_name)).alias("shift_1"),
        col(column_name).shift(-1).lt(col(column_name)).alias("shift_neg_1"),
    ])
    .with_column(col("shift_1").and(col("shift_neg_1")).alias("diff"))
    .filter(col("diff"));

    lf.left_join(maxima_lf, col("date"), col("date"), None)
}

let dts = vec![
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 0, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 1, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 2, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 3, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 4, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 5, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 6, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 7, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 8, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 9, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 10, 0),
    NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 11, 0),
];
let data = vec![2., 1., 1., 1., 1., 1., 1., 1., 1., 3., 5., 6.];
let series = Series::new("data", data);
let df = DataFrame::new(vec![
    Date64Chunked::new_from_naive_datetime("date", &*dts).into(),
    series
])?;
let res = foo(df.lazy(), "data").collect()?;
```

#### What is the actual behavior?

Panics with 'implementation error, cannot get ref Int64 from Date64'

#### What is the expected behavior?

Don't panic :)

**Comment by TheDan64** [comment]
Features are `["lazy", "random", "temporal", "dtype-date64", "dtype-u64", "object"]`

**Comment by ritchie46** [comment]
Thanks, this should be fixed by #574.
